File: js/net/src/ietf/namespace_count.ts
import { MSG_PARAM_NAMESPACE_COUNT, type Parameters, SetupOption, type SetupOptions } from "./parameters.ts";
import { type IetfVersion, Version } from "./version.ts";

/**
 * The MoQ Namespace Count extension (draft-lcurley-moq-namespace-count-00).
 *
 * moq-transport answers a SUBSCRIBE_NAMESPACE with a REQUEST_OK and then a NAMESPACE per
 * matching namespace, with nothing separating the ones the publisher already had from the
 * ones that showed up a moment later. A subscriber can see that a namespace is present,
 * but never that it is absent.
 *
 * This extension puts the size of that initial set in the response, so the set is complete
 * once that many NAMESPACE messages have arrived. It is moq-lite's `ANNOUNCE_OK.Active
 * Count` on the moq-transport wire.
 *
 * Unknown Message Parameters close the session, so unlike the MoQ Solicit extension this
 * one has to be negotiated: the count only goes to a peer whose SETUP asked for it.
 *
 * @module
 * @internal
 */

/**
 * Whether this version carries the initial set on the SUBSCRIBE_NAMESPACE response stream
 * at all.
 *
 * Draft-14/15 answer with PUBLISH_NAMESPACE requests on streams of their own and have no
 * NAMESPACE message, so there is nothing on the response stream to count.
 *
 * @internal
 */
export function namespaceCountSupported(version: IetfVersion): boolean {
	return version !== Version.DRAFT_14 && version !== Version.DRAFT_15;
}

/**
 * Whether the peer asked us to report the count.
 *
 * Absent and 0 are the same statement: unlike the MoQ Solicit extension, nothing here
 * changes based on whether a peer that wants nothing implements this.
 *
 * @internal
 */
export function namespaceCountFromSetup(params: SetupOptions, version: IetfVersion): boolean {
	if (!namespaceCountSupported(version)) return false;
	const value = params.getVarint(SetupOption.NamespaceCount);
	return value !== undefined && value !== 0n;
}

/**
 * Ask the peer to report the count.
 *
 * Unconditional on every version that can answer, since we open a SUBSCRIBE_NAMESPACE for
 * every prefix we want and each one wants a boundary.
 *
 * @internal
 */
export function namespaceCountIntoSetup(params: SetupOptions, version: IetfVersion) {
	if (!namespaceCountSupported(version)) return;
	params.setVarint(SetupOption.NamespaceCount, 1n);
}

/**
 * The initial set size a SUBSCRIBE_NAMESPACE response reported, or `undefined` on a peer
 * that does not implement the extension. `0n` is a real answer: the prefix is empty.
 *
 * @internal
 */
export function namespaceCountFromResponse(params: Parameters): bigint | undefined {
	return params.vars.get(MSG_PARAM_NAMESPACE_COUNT);
}

/**
 * Report the initial set size on a SUBSCRIBE_NAMESPACE response.
 *
 * @internal
 */
export function namespaceCountIntoResponse(params: Parameters, count: bigint) {
	params.vars.set(MSG_PARAM_NAMESPACE_COUNT, count);
}

File: rs/moq-net/src/ietf/namespace_count.rs
//! The MoQ Namespace Count extension (draft-lcurley-moq-namespace-count-00).
//!
//! Nothing in moq-transport says where the NAMESPACE messages answering a
//! SUBSCRIBE_NAMESPACE stop replaying what already existed and start reporting changes,
//! so a subscriber can only guess from the stream going quiet. When both endpoints
//! declare the NAMESPACE_COUNT option, the publisher puts the size of that initial set on
//! its REQUEST_OK, the way moq-lite's ANNOUNCE_OK does.

use super::Version;

/// NAMESPACE_COUNT Setup Option: the sender counts its initial NAMESPACE messages and
/// reads the peer's count. Even, so the value is a bare varint.
pub const NAMESPACE_COUNT: u64 = 0x40B64;

/// NAMESPACE_COUNT REQUEST_OK parameter: how many NAMESPACE messages immediately follow.
/// Even, so the value is a bare varint.
pub const NAMESPACE_COUNT_PARAM: u64 = 0x40B66;

/// Whether this version answers SUBSCRIBE_NAMESPACE with NAMESPACE messages. Draft-14/15
/// answer with PUBLISH_NAMESPACE requests instead, which the stream cannot count.
fn supported(version: Version) -> bool {
	!matches!(version, Version::Draft14 | Version::Draft15)
}

/// Whether the peer declared the option. We always declare it, so this alone decides
/// whether the extension is negotiated.
pub fn from_setup(params: &super::Parameters, version: Version) -> bool {
	supported(version) && params.get_varint(super::ParameterVarInt::NamespaceCount).is_some()
}

/// Declare that we count our initial set and read the peer's.
pub fn into_setup(params: &mut super::Parameters, version: Version) {
	if supported(version) {
		params.set_varint(super::ParameterVarInt::NamespaceCount, 1);
	}
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn only_versions_with_namespace_negotiate() {
		for version in [Version::Draft14, Version::Draft15, Version::Draft16, Version::Draft22] {
			let mut params = super::super::Parameters::default();
			assert!(!from_setup(&params, version));
			into_setup(&mut params, version);
			assert_eq!(from_setup(&params, version), supported(version), "{version:?}");
		}
	}
}
